specification/sections/03-evaluation-context.md

```diff
@@ -42,15 +42,39 @@ see: [structure](../types.md#structure), [datetime](../types.md#datetime)
 
 The key uniquely identifies a field in the `evaluation context` and it should be unique across all types to avoid any collision when marshalling the `evaluation context` by the provider.
 
-### 3.2 Merging Context
+### 3.2 Context levels and merging
 
-#### Requirement 3.2.1
+#### Condition 3.2.1
+
+> The implementation uses the static-context paradigm.
+
+see: [static-context paradigm](../glossary.md#static-context-paradigm)
+
+##### Conditional Requirement 3.2.1.1
+
+> The API **MUST** have a method for supplying `evaluation context`.
+
+API (global) `evaluation context` can be used to supply data to flag evaluation, such as (but not limited to) user name, email, or shopping cart items.
+
+##### Conditional Requirement 3.2.1.2
+
+> The Client and invocation **MUST NOT** have a method for supplying `evaluation context`.
+
+In the static-context paradigm, context is global. The client and invocation cannot supply evaluation context. 
```

Comment:
Interesting. What's the benefit of enforcing that the invocation cannot supply a context? I'm struggling to see the use-case where this is beneficial to the user.

I can understand that you could get the state where two contexts contain different values for the same key in the context, but in my mind it's trivial to say that the invocation-context overrides the global-context for situations where we have a key-clash.

Given that CloudBees Feature Management allows setting a global context and invocation context, it basically means that CBFM will always operate in the evaluation context paradigm. I'm not sure if I'm understanding this correctly though, or if this is an issue.

Reply:
As you note, some providers (Cloudbees being one of them) do local evaluation in client apps, and therefore can do "dynamic" evaluation (involving an invocation context) without issue. This isn't the case with other vendors. Those SDKs basically have a "bulk evaluate + cache + return-from-cache" paradigm, where a change in context triggers another bulk evaluation (rules are never sent to the client).

Allowing dynamic context in this case is basically a no-go, it's a huge performance footgun. While it's possible for Cloudbees to fit into the static-context paradigm, it's not possible for the bulk-evaluated SDKs to fit into a dynamic context paradigm. We're trying to walk a fine-line here to find an abstraction that works across as many paradigms as possible.

@moredip  did a really good job explaining this in his recent blog post: https://docs.openfeature.dev/blog/catering-to-the-client-side

I'm definitely interested in getting more thoughts from you on this, since Cloudbees is one of the two vendors I'm aware of whose client SDKs works by shipping a ruleset.

Reply:
This [OFEP](https://github.com/open-feature/ofep/pull/41) is also highly relevant.